From 04de718435094a6e37a19b46da16e97acc76fd53 Mon Sep 17 00:00:00 2001
From: Gustavo Duarte <gduarte@redhat.com>
Date: Wed, 19 Nov 2014 23:45:11 -0500
Subject: [scsi] ipr: don't log error messages when applications issues illegal requests

Message-id: <1416440711-16666-1-git-send-email-gduarte@redhat.com>
Patchwork-id: 100381
O-Subject: [RHEL7.1 PATCH BZ 1163019] ipr: don't log error messages when applications issues illegal requests
Bugzilla: 1163019
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Don Zickus <dzickus@redhat.com>

BZ 1163019
Upstream Status: 3185ea63907cb281cfdc5aa29aa2d855826ad16a
Build Info: http://brewweb.devel.redhat.com/brew/taskinfo?taskID=8269576
ppc64le: http://brewweb.devel.redhat.com/brew/taskinfo?taskID=8269584
Tested: Kernel boots fine on ppc64 and ppc64le.

commit 3185ea63907cb281cfdc5aa29aa2d855826ad16a
Author: wenxiong@linux.vnet.ibm.com <wenxiong@linux.vnet.ibm.com>
Date:   Wed Sep 24 16:25:47 2014 -0500

    ipr: don't log error messages when applications issues illegal requests

    Failing Device information are logged when IOA firmware detected these
    illegal request such as IOA firmware doesn't support inquiry with page
    code 2. The patch fixes the issue.

    Signed-off-by: Brian King <brking@linux.vnet.ibm.com>
    Tested-by: Wen Xiong <wenxiong@linux.vnet.ibm.com>
    Signed-off-by: Christoph Hellwig <hch@lst.de>

This also merges the follow-up typo fix, commit d247a70a4d67623eb738176ee703a7bb8b6c325c

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/ipr.c b/drivers/scsi/ipr.c
index 4ee3d8a..c788853 100644
--- a/drivers/scsi/ipr.c
+++ b/drivers/scsi/ipr.c
@@ -2443,6 +2443,7 @@ static void ipr_handle_log_data(struct ipr_ioa_cfg *ioa_cfg,
 {
  u32 ioasc;
  int error_index;
+ struct ipr_hostrcb_type_21_error *error;
 
  if (hostrcb->hcam.notify_type != IPR_HOST_RCB_NOTIF_TYPE_ERROR_LOG_ENTRY)
   return;
@@ -2467,6 +2468,15 @@ static void ipr_handle_log_data(struct ipr_ioa_cfg *ioa_cfg,
  if (!ipr_error_table[error_index].log_hcam)
   return;
 
+ if (ioasc == IPR_IOASC_HW_CMD_FAILED &&
+     hostrcb->hcam.overlay_id == IPR_HOST_RCB_OVERLAY_ID_21) {
+  error = &hostrcb->hcam.u.error64.u.type_21_error;
+
+  if (((be32_to_cpu(error->sense_data[0]) & 0x0000ff00) >> 8) == ILLEGAL_REQUEST &&
+   ioa_cfg->log_level <= IPR_DEFAULT_LOG_LEVEL)
+    return;
+ }
+
  ipr_hcam_err(hostrcb, "%s\n", ipr_error_table[error_index].error);
 
  /* Set indication we have logged an error */
diff --git a/drivers/scsi/ipr.h b/drivers/scsi/ipr.h
index c24c747..fa82c00 100644
--- a/drivers/scsi/ipr.h
+++ b/drivers/scsi/ipr.h
@@ -130,6 +130,7 @@
 #define IPR_IOASC_HW_DEV_BUS_STATUS   0x04448500
 #define IPR_IOASC_IOASC_MASK   0xFFFFFF00
 #define IPR_IOASC_SCSI_STATUS_MASK  0x000000FF
+#define IPR_IOASC_HW_CMD_FAILED   0x046E0000
 #define IPR_IOASC_IR_INVALID_REQ_TYPE_OR_PKT 0x05240000
 #define IPR_IOASC_IR_RESOURCE_HANDLE  0x05250000
 #define IPR_IOASC_IR_NO_CMDS_TO_2ND_IOA  0x05258100
-- 
1.7.1